File: Python/Calendar.py
import tkinter as tk
from tkinter import scrolledtext
import calendar
from datetime import datetime
import json
import os

class CalendarApp:
    def __init__(self, root):
        self.root = root
        self.root.title("メモ・カレンダー")
        self.root.geometry("800x600")
        
        self.curr_date = datetime.now()
        self.data_file = "calendar_data.json"
        self.memo_data = self.load_data()
        
        self.create_widgets()
        self.draw_calendar()

    def load_data(self):
        if os.path.exists(self.data_file):
            with open(self.data_file, 'r', encoding='utf-8') as f:
                return json.load(f)
        return {}

    def save_data(self, key, text):
        self.memo_data[key] = text
        with open(self.data_file, 'w', encoding='utf-8') as f:
            json.dump(self.memo_data, f, ensure_ascii=False)

    def create_widgets(self):
        self.header = tk.Frame(self.root)
        self.header.pack(pady=10)
        
        tk.Button(self.header, text="◀ 前月", command=lambda: self.change_month(-1)).pack(side=tk.LEFT)
        self.month_label = tk.Label(self.header, text="", font=("Arial", 16, "bold"))
        self.month_label.pack(side=tk.LEFT, padx=20)
        tk.Button(self.header, text="次月 ▶", command=lambda: self.change_month(1)).pack(side=tk.LEFT)

        self.cal_frame = tk.Frame(self.root)
        self.cal_frame.pack(expand=True, fill="both", padx=10, pady=10)

    def draw_calendar(self):
        for widget in self.cal_frame.winfo_children():
            widget.destroy()

        y, m = self.curr_date.year, self.curr_date.month
        self.month_label.config(text=f"{y}年 {m}月")

        # 曜日ヘッダー
        days = ["日", "月", "火", "水", "木", "金", "土"]
        for i, d in enumerate(days):
            color = "red" if i == 0 else "blue" if i == 6 else "black"
            tk.Label(self.cal_frame, text=d, fg=color, font=("Arial", 10, "bold")).grid(row=0, column=i, sticky="nsew")

        cal = calendar.Calendar(firstweekday=6)
        weeks = cal.monthdayscalendar(y, m)

        for r, week in enumerate(weeks):
            for c, day in enumerate(week):
                if day == 0: continue
                
                cell = tk.Frame(self.cal_frame, bd=1, relief="sunken")
                cell.grid(row=r+1, column=c, sticky="nsew")
                
                # 数字の色分け
                num_color = "red" if c == 0 else "blue" if c == 6 else "black"
                tk.Label(cell, text=day, fg=num_color, font=("Arial", 10)).pack(anchor="nw")

                # メモ入力欄
                key = f"{y}-{m}-{day}"
                txt = scrolledtext.ScrolledText(cell, height=3, width=10, font=("Arial", 8))
                txt.insert(tk.END, self.memo_data.get(key, ""))
                txt.pack(expand=True, fill="both")
                
                # 入力されたら自動保存
                txt.bind("<KeyRelease>", lambda e, k=key, t=txt: self.save_data(k, t.get("1.0", tk.END).strip()))

        for i in range(7): self.cal_frame.grid_columnconfigure(i, weight=1)
        for i in range(len(weeks)+1): self.cal_frame.grid_rowconfigure(i, weight=1)

    def change_month(self, diff):
        new_month = self.curr_date.month + diff
        new_year = self.curr_date.year
        if new_month > 12:
            new_month = 1
            new_year += 1
        elif new_month < 1:
            new_month = 12
            new_year -= 1
        self.curr_date = datetime(new_year, new_month, 1)
        self.draw_calendar()

if __name__ == "__main__":
    root = tk.Tk()
    app = CalendarApp(root)
    root.mainloop()

File: csharp/Calendar.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Text.Json;
using System.Windows.Forms;

namespace MyCalendarApp
{
    public partial class Form1 : Form
    {
        private DateTime currentViewDate = DateTime.Now;
        private TableLayoutPanel calendarGrid;
        private Label monthYearLabel;
        private Dictionary<string, string> memoData = new Dictionary<string, string>();
        private string filePath = "calendar_data.json";
        private int expandedCol = -1; // 拡張する列（曜日）のインデックス

        public Form1()
        {
            this.Text = "C# メモカレンダー";
            this.Size = new Size(900, 700);
            LoadData();
            InitializeLayout();
            RenderCalendar();
        }

        private void InitializeLayout()
        {
            var rootPanel = new DockLayoutPanel(); // 全体レイアウト
            
            // ヘッダー（月移動）
            var header = new Panel { Dock = DockStyle.Top, Height = 60, BackColor = Color.FromArgb(45, 45, 48) };
            monthYearLabel = new Label { ForeColor = Color.White, Font = new Font("Arial", 16, FontStyle.Bold), AutoSize = true, Location = new Point(350, 15) };
            
            var btnPrev = new Button { Text = "◀ 前月", Location = new Point(250, 15), FlatStyle = FlatStyle.Flat, ForeColor = Color.White };
            btnPrev.Click += (s, e) => { ChangeMonth(-1); };
            
            var btnNext = new Button { Text = "次月 ▶", Location = new Point(550, 15), FlatStyle = FlatStyle.Flat, ForeColor = Color.White };
            btnNext.Click += (s, e) => { ChangeMonth(1); };

            header.Controls.Add(monthYearLabel);
            header.Controls.Add(btnPrev);
            header.Controls.Add(btnNext);

            // カレンダーグリッド
            calendarGrid = new TableLayoutPanel { Dock = DockStyle.Fill, ColumnCount = 7, RowCount = 7 };
            for (int i = 0; i < 7; i++) calendarGrid.ColumnStyles.Add(new ColumnStyle(SizeType.Percent, 14.28f));
            
            this.Controls.Add(calendarGrid);
            this.Controls.Add(header);
        }

        private void RenderCalendar()
        {
            calendarGrid.Controls.Clear();
            calendarGrid.RowStyles.Clear();
            calendarGrid.RowStyles.Add(new RowStyle(SizeType.Absolute, 40)); // 曜日ヘッダー行

            monthYearLabel.Text = $"{currentViewDate.Year}年 {currentViewDate.Month}月";

            // 曜日ヘッダー作成
            string[] weekDays = { "日", "月", "火", "水", "木", "金", "土" };
            for (int i = 0; i < 7; i++)
            {
                var lbl = new Label { Text = weekDays[i], Dock = DockStyle.Fill, TextAlign = ContentAlignment.MiddleCenter, Font = new Font("Arial", 10, FontStyle.Bold) };
                if (i == 0) lbl.ForeColor = Color.Red;
                if (i == 6) lbl.ForeColor = Color.Blue;
                calendarGrid.Controls.Add(lbl, i, 0);
            }

            DateTime firstDay = new DateTime(currentViewDate.Year, currentViewDate.Month, 1);
            int daysInMonth = DateTime.DaysInMonth(currentViewDate.Year, currentViewDate.Month);
            int startOffset = (int)firstDay.DayOfWeek;

            int row = 1;
            for (int i = 0; i < 42; i++)
            {
                int dayNum = i - startOffset + 1;
                int col = i % 7;
                if (col == 0 && i > 0) row++;
                if (row > 6) break;

                var cell = new Panel { Dock = DockStyle.Fill, BorderStyle = BorderStyle.FixedSingle, BackColor = Color.White };
                
                // 列拡張の背景色
                if (col == expandedCol) cell.BackColor = Color.LightYellow;

                if (dayNum > 0 && dayNum <= daysInMonth)
                {
                    string dateKey = $"{currentViewDate.Year}-{currentViewDate.Month}-{dayNum}";

                    // 数字ラベル（クリックで列拡張）
                    var lbl = new Label { Text = dayNum.ToString(), Dock = DockStyle.Top, Cursor = Cursors.Hand, Font = new Font("Arial", 9, FontStyle.Bold) };
                    if (col == 0) lbl.ForeColor = Color.Red;
                    if (col == 6) lbl.ForeColor = Color.Blue;
                    lbl.Click += (s, e) => { ToggleColumn(col); };

                    // メモ入力欄
                    var txt = new TextBox { Multiline = true, Dock = DockStyle.Fill, BorderStyle = BorderStyle.None, Font = new Font("MS UI Gothic", 9) };
                    txt.Text = memoData.ContainsKey(dateKey) ? memoData[dateKey] : "";
                    txt.TextChanged += (s, e) => { memoData[dateKey] = txt.Text; SaveData(); };

                    cell.Controls.Add(txt);
                    cell.Controls.Add(lbl);
                }
                calendarGrid.Controls.Add(cell, col, row);
            }

            // 行の高さ調整（拡張時）
            for (int r = 1; r <= 6; r++) calendarGrid.RowStyles.Add(new RowStyle(SizeType.Percent, 16.6f));
        }

        private void ToggleColumn(int colIndex)
        {
            expandedCol = (expandedCol == colIndex) ? -1 : colIndex;
            // 列の幅を動的に変更
            calendarGrid.ColumnStyles.Clear();
            for (int i = 0; i < 7; i++)
            {
                float width = (expandedCol == -1) ? 14.28f : (i == expandedCol ? 40f : 10f);
                calendarGrid.ColumnStyles.Add(new ColumnStyle(SizeType.Percent, width));
            }
            RenderCalendar();
        }

        private void ChangeMonth(int diff) { currentViewDate = currentViewDate.AddMonths(diff); expandedCol = -1; RenderCalendar(); }

        private void SaveData() { File.WriteAllText(filePath, JsonSerializer.Serialize(memoData)); }
        private void LoadData() { if (File.Exists(filePath)) memoData = JsonSerializer.Deserialize<Dictionary<string, string>>(File.ReadAllText(filePath)); }
    }

    // レイアウト調整用ヘルパー
    public class DockLayoutPanel : Panel { public DockLayoutPanel() { this.Dock = DockStyle.Fill; } }
}
